### ISSUE TITLE
[DOC-META] Migration and upgrade documentation

### ISSUE THREAD

**Issue by setiah** [opensearch-project/documentation-website#745]
## Description

This is a Meta issue to track any gaps in documentation for Migrations to OpenSearch. Since Migrations is a broad topic, this meta issue would be helpful in capturing and linking back all migration related documentation pieces together at one place.

The scope of Migration includes but is not limited to - 
* Migration from previous Elasticsearch versions to any OpenSearch version
* Migration from ODFE to OpenSearch
* Migration from a lower OpenSearch version to higher OpenSearch version
* Migration on any platform (RPM, Windows, Docker, etc.)
* Preferred tools/mechanisms available for migration

## Sub-issues 

Will add more later.. 

- [ ] Describe all data migration mechanisms with their pros/cons and version support matrix (from-to version)
- [ ] Migration from ODFE and Elasticsearch to OpenSearch.
- [ ] Migration from Elasticsearch Clients to OpenSearch Clients
- [ ] Platform specific considerations e.g. Docker, RPM, Windows etc.
- [ ] Pre-migration checks - best practices.
- [ ] Migrating static configurations
- [ ] Migrating ODFE plugins to OpenSearch

This issue is open for feedback/suggestions. We'll keep adding more stuff in here as we identify more gaps.

Note (added by JeffH-AWS) - a blog post already exists for migrating from ODFE to OpenSearch https://opensearch.org/blog/technical-posts/2021/07/how-to-upgrade-from-opendistro-to-opensearch/

**Comment by Naarcha-AWS** [comment]
@setiah: Do we have any resources to help us get started on the migrations you identified, or a POC who could give one of the writers a demo of how it works?

**Comment by setiah** [comment]
Thanks @Naarcha-AWS for picking this up. There are a couple of good blogposts resources that should help you get started. These have a lot of details on  ODFE to OpenSearch and Elasticsearch to OpenSearch migrations. We don't have a dedicated person on migrations yet but if you get stuck, I can help or we can seek help from OpenSearch team by creating an issue on OpenSearch/respective plugin repo to get the answers.   

* https://opensearch.org/blog/technical-posts/2021/07/how-to-upgrade-from-opendistro-to-opensearch/
* https://opensearch.org/blog/technical-posts/2021/10/moving-from-opensource-elasticsearch-to-opensearch/
* https://eliatra.com/blog/elasticsearch-to-opensearch-migration-snapshot-restore/

**Comment by JeffH-AWS** [comment]
I'm currently going through the exercise of standing up a dev environment so I can validate installation steps and recommendations as I go.  Planning to start digging in to this in depth next week.

**Comment by hdhalter** [comment]
Hi @JeffH-AWS , the OpenDistro project has been archived and we've been asked to update the OpenDistro documentation to remove all download links. In addition, we need to update the OpenSearch documentation to include information on migrating from OpenDistro to OpenSearch, as we are recommending that anyone using OpenDistro move to OpenSearch. The instructions for migrating are here: https://opensearch.org/blog/technical-posts/2021/07/how-to-upgrade-from-opendistro-to-opensearch/. We refer to the migration here: https://opensearch.org/docs/latest/upgrade-to/index/. You can probably point to the blog, but should confirm with Minal that the migration instructions are up to date. Is it possible for you to prioritize this week on the migration part? I can see if someone else has bandwidth to update the OpenDistro doc. Thanks!

**Comment by JeffH-AWS** [comment]
No problem.  Taking a look.

**Comment by JeffH-AWS** [comment]
@hdhalter can I please have clarification on `we've been asked to update the OpenDistro documentation to remove all download links`?  Where do these links live currently and what we we pointing them to?  Or should they just be removed?

**Comment by JeffH-AWS** [comment]
Requested feedback from Sarat on the blog post relating to version compatibility for the process outlines.

**Comment by JeffH-AWS** [comment]
Will add to https://opensearch.org/docs/latest/upgrade-to/index/

**Comment by JeffH-AWS** [comment]
@cwillum will be reviewing the download links that need to be removed/updated.

**Comment by hdhalter** [comment]
Hi Jeff, I opened a new issue for updating the Open Distro documentation and assigned to @cwillum. https://github.com/opensearch-project/documentation-website/issues/840.

**Comment by JeffH-AWS** [comment]
Meeting with @setiah this morning to plan this out.

**Comment by JeffH-AWS** [comment]
Circling back - this is fully in progress now, but the p0 is Docker major/minor version upgrades. I've started data gathering and have built a lab environment to use for testing, including separate `docker-compose.yml` files for each minor version of `1.x` and `2.x`. I have compiled a list of questions that require feedback from the devs and PM.

Additionally, the table of contents for the page is being restructured. Right now they are redoing the sections relating to installation and cluster configuration. I will wait until that's done to open a draft PR for this to avoid needing to rebase later.

**Comment by JeffH-AWS** [comment]
Currently blocked on this while I wait to hear back from developer resources. I've compiled a list of questions. Here's a list of open questions:

```
Can we assume that a user performing an upgrade will remain on the same infra (number of nodes, network paths, volumes, etc.)? Or should we be writing this as if there is no guarantee that a user will recycle the same hardware? Should we add any prefaces on this topic?
*Answer:* 


For Docker minor version upgrades is it literally as easy as just changing the image that it points to since data is persistent and there are no breaking changes within a single major version’s releases?
*Answer:*


Which methods are available for upgrading a Docker cluster? According to https://opensearch.org/docs/latest/upgrade-to/index/ the methods are snapshot & restore, restart upgrade, rolling upgrade, node replacement, and the “upgrade tool.” Is that list correct and exhaustive?
*Answer:*


Snapshot & restore is described as follows: “Snapshots are backups of a cluster’s indexes and state. State includes cluster settings, node information, index metadata (mappings, settings, templates, etc.), and shard allocation.” How does this work in practice? Is there a 1:1 relationship between nodes in the cluster when the snapshot was taken and how many nodes need to exist afterwards?
*Answer:*


Do we have a quick “check” that can be performed to ensure that everything is intact after an upgrade? I.e., does just running curl "https://localhost:9200/_cluster/health?pretty=true" or something along those lines give us a definitive answer or are there other things a user should check?
*Answer:*

Are there any kind of system architecture limitations when it comes to snapshot & restore, or is a snapshot completely portable? In other words, are there any settings stored in a snapshot that might be environment-specific, such as JDK options, etc.
*Answer:*


Restart upgrade or rolling upgrade - these sound like two different methods. Are they? Or is it just two strategies for implementing the same process? And, if so, is one “better” than the other?
*Answer:*


How does node replacement differ from rolling upgrade?
*Answer:*


What is the “[upgrade tool](https://opensearch.org/docs/latest/upgrade-to/upgrade-to/#upgrade-tool) (https://opensearch.org/docs/latest/upgrade-to/upgrade-to/#upgrade-tool)” and is it useful for OS→OS major/minor upgrades as well? Or is only useful when moving from ES to OS?
*Answer:*


Which methods allow for disaster recovery/rollback? For example, if you are running OS 1.0 and then upgrade to 1.3.x, let’s say you do it with a snapshot. If you keep the snapshot from 1.0 stored safely, you could restore back to that point in time, but is there a strategy for retaining the delta between docs that were indexed prior to the 1.0 snapshot and docs that were indexed after upgrading to 1.3.x, but before deciding to roll it back?
*Answer:*
```

**Comment by JeffH-AWS** [comment]
Engineers have been engaged to assist with the questions. I'm also waiting on an update to the ToC for the installation and cluster config sections before I open up a draft PR. The ToC update is supposed to be merged today.

**Comment by JeffH-AWS** [comment]
Docker minor version doc is pretty straightforward - take a snapshot so that the node can be restored to its original condition if needed.

Open question - is there a way to account for the diff in indexed data between when a snapshot was taken and when the rollback happens? I.e., stuff that was indexed on the new version before the decision is made to roll back.

I've currently got a merry-go-round of questions I'm serving to SMEs, but long response times are slowing my progress. In the meantime, I've stood up a remote environment that I can use to validate some of the steps, but I need to learn how reindexing and snapshots work, and both of those topics have subcategories that need explored.

**Comment by JeffH-AWS** [comment]
Request for engineering assistance from the migrations team: https://github.com/opensearch-project/opensearch-migrations/issues/34

**Comment by JeffH-AWS** [comment]
Current status -

An internal resource was identified (part of migrations team) that can help with engineering-specific questions about the product.

I'm currently mapping assets that can be backed up and restored to the new cluster. In my original cluster I indexed some sample data, created a new user, and modified settings in Dashboards. So far I'm only able to migrate the sample data and the user. I'm still working out how to preserve the cluster's global state, [which includes](https://opensearch.org/docs/latest/api-reference/snapshots/restore-snapshot/):

- Persistent cluster settings
- Index templates
- Legacy index templates
- Ingest pipelines
- Index lifecycle policies

After each attempt I have to wipe the new cluster because any tests against a partially migrated cluster aren't representative of an actual migration, and it is much easier to work with a single variable at a time. For example, if my `ecommerce` index is restored, then my future commands will be invalid unless I modify them for one-offs, which isn't efficient.

**Comment by JeffH-AWS** [comment]
Waiting for confirmation from a SME for the following:

- Cluster security settings (opensearch-security)
    - Since the index can't be migrated, these are backed up by invoking securityadmin.sh -backup, and then after launching the new cluster dropping those YAMLs into a directory and running securityadmin.sh -cd /path/to/directory
- TLS certificates
    - Must be backed up manually
- Indexes
    - Snapshot & Restore to new cluster (only works on non-system indexes based on my testing)
    - Remote reindexing
        - Seems like remote indexing provides more flexibility in maintaining cluster availability?
- Cluster state
    - Supposedly snapshot & restore "saves the cluster state" but there's a huge disclaimer in our docs that you can't restore a snapshot that contains the cluster state. My testing also resulted in snapshot restore operations failing with a very unhelpful response unless I set include global state to false.
    - Includes (must be backed up manually unless someone has a better way):
        - opensearch.yml
        - opensearch_dashboards.yml
        - jvm.options
        - log4j2.properties
        - I'm including this one because the [ES docs call it out](https://www.elastic.co/guide/en/elasticsearch/reference/current/settings.html) as one of the "three configuration files."

If there's anything I'm missing please let me know. In terms of my research and testing, this is one of the biggest "I don't know what I don't know" subjects so I'd like to get some feedback from the brains in this channel please!

**Comment by JeffH-AWS** [comment]
I have still not gotten any information about the previous questions - should get follow up this week.

In the meantime, I'm trying to work out a process for snapshot restoration. I wonder if it's possible to launch the new cluster without security enabled, migrate, then enable the security plugin -- that might help mitigate the problem with the global cluster state not being able to get restored.

I'm also running into more very unhelpful errors while attempting to restore snapshots. My latest snapshot restore operation caused my Dashboards node to break.

```
{"type":"error","@timestamp":"2022-12-12T22:13:30Z","tags":[],"pid":1,"level":"error","error":{"message":"Internal Server Error","name":"Error","stack":"Error: Internal Server Error
    at HapiResponseAdapter.toInternalError (/usr/share/opensearch-dashboards/src/core/server/http/router/response_adapter.js:80:19)
    at Router.handle (/usr/share/opensearch-dashboards/src/core/server/http/router/router.js:177:34)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at handler (/usr/share/opensearch-dashboards/src/core/server/http/router/router.js:124:50)
    at exports.Manager.execute (/usr/share/opensearch-dashboards/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at Object.internals.handler (/usr/share/opensearch-dashboards/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at exports.execute (/usr/share/opensearch-dashboards/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at Request._lifecycle (/usr/share/opensearch-dashboards/node_modules/@hapi/hapi/lib/request.js:371:32)
    at Request._execute (/usr/share/opensearch-dashboards/node_modules/@hapi/hapi/lib/request.js:281:9)"},"url":"http://ec2-52-41-94-89.us-west-2.compute.amazonaws.com:5601/","message":"Internal Server Error"}
{"type":"response","@timestamp":"2022-12-12T22:13:30Z","tags":[],"pid":1,"method":"get","statusCode":500,"req":{"url":"/","method":"get","headers":{"host":"ec2-52-41-94-89.us-west-2.compute.amazonaws.com:5601","connection":"keep-alive","cache-control":"max-age=0","upgrade-insecure-requests":"1","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9","accept-encoding":"gzip, deflate","accept-language":"en-US,en;q=0.9","securitytenant":""},"remoteAddress":"50.39.252.121","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"},"res":{"statusCode":500,"responseTime":11,"contentLength":9},"message":"GET / 500 11ms - 9.0B"}
```

**Comment by JeffH-AWS** [comment]
Upgrade overview and Docker upgrade steps are coming along nicely now. I had a lot of difficulty planning the documents because I wanted to organize the content in such a way that it minimizes repetition of content, but that it doesn't oversimplify any of the steps like a lot of other documentation - like telling users to go do X, Y, and Z when none of those processes are clearly documented anywhere.

The primary overview page will touch on important concepts that should be considered before upgrade. Tolerance for downtime, reviewing aspects of the cluster for compatibility with the target version, etc.

Each individual page will address specific steps for the distribution/platform, but will rely on the concepts introduced in the overview page. That way I don't need to re-explain what a rolling upgrade is a half dozen different ways.

**Comment by JeffH-AWS** [comment]
Upgrade overview and Docker upgrade steps are currently in technical review.

**Comment by JeffH-AWS** [comment]
When the tech review is completed (might be a little slow this week since a lot of folks are offline) I'm going to migrate these docs to a new branch cut from main with the ToC changes so that I don't have to deal with the headache of rebasing.

**Comment by JeffH-AWS** [comment]
Open https://github.com/opensearch-project/OpenSearch/issues/5738 for a bug I found with env variables in 1.3.7

**Comment by JeffH-AWS** [comment]
One of the asks from the technical review of my upgrade overview page is to add a version compatibility matrix for Lucene indexes.

In the official [Lucene documentation](https://cwiki.apache.org/confluence/display/LUCENE/BackwardsCompatibility), they state:

```
File Formats
File formats are back-compatible between major versions. Version X.N should be able to read indexes generated by any version after and including version X-1.0, but may-or-may-not be able to read indexes generated by version X-2.N.

Note: Older releases are never guaranteed to be able to read indexes generated by newer releases. When this is attempted, a predictable error should be generated.
```

Since we don't have an OpenSearch-native upgrade path that spans N+2 major versions, I'm going to test this by running a Docker instance with each GA version of OpenSearch.

I'll also go back one major version (to 6.8, I think?) of ElasticSearch to check from there to the fork.

**Comment by JeffH-AWS** [comment]
Replacing this issue with https://github.com/opensearch-project/documentation-website/issues/2830 so we can start from a fresh issue since we'll be looking for community suggestions and feedback to guide our upcoming docs work.
